NExtends/Expressions/ExpressionEqualityComparer.cs

```diff
@@ -0,0 +1,239 @@
+			}
+			if (x is MemberInitExpression)
+			{
+				var mix = (MemberInitExpression)x;
+				var miy = (MemberInitExpression)y;
+				return ExpressionsEqual(mix.NewExpression, miy.NewExpression, rootX, rootY)
+					   && MemberInitsEqual(mix.Bindings, miy.Bindings, rootX, rootY);
+			}
+			if (x is NewArrayExpression)
+			{
+				var nx = (NewArrayExpression)x;
+				var ny = (NewArrayExpression)y;
+				return CollectionsEqual(nx.Expressions, ny.Expressions, rootX, rootY);
+			}
+			if (x is NewExpression)
+			{
+				var nx = (NewExpression)x;
+				var ny = (NewExpression)y;
+				return
+					Equals(nx.Constructor, ny.Constructor)
+					&& CollectionsEqual(nx.Arguments, ny.Arguments, rootX, rootY)
+					&& (nx.Members == null && ny.Members == null
+						|| nx.Members != null && ny.Members != null && CollectionsEqual(nx.Members, ny.Members));
+			}
+			if (x is ConditionalExpression)
+			{
+				var cx = (ConditionalExpression)x;
+				var cy = (ConditionalExpression)y;
+				return
+					ExpressionsEqual(cx.Test, cy.Test, rootX, rootY)
+					&& ExpressionsEqual(cx.IfFalse, cy.IfFalse, rootX, rootY)
+					&& ExpressionsEqual(cx.IfTrue, cy.IfTrue, rootX, rootY);
+			}
+
+			throw new NotImplementedException(x.ToString());
+		}
+
+		private static Boolean IsAnonymousType(Type type)
+		{
+			Boolean hasCompilerGeneratedAttribute = type.GetCustomAttributes(typeof(CompilerGeneratedAttribute), false).Any();
+			Boolean nameContainsAnonymousType = type.FullName.Contains("AnonymousType");
+			Boolean isAnonymousType = hasCompilerGeneratedAttribute && nameContainsAnonymousType;
+
+			return isAnonymousType;
+		}
+
+		private static bool MemberInitsEqual(ICollection<MemberBinding> bx, ICollection<MemberBinding> by, LambdaExpression rootX, LambdaExpression rootY)
+		{
+			if (bx.Count != by.Count)
+			{
+				return false;
+			}
+
+			if (bx.Concat(by).Any(b => b.BindingType != MemberBindingType.Assignment))
+				throw new NotImplementedException("Only MemberBindingType.Assignment is supported");
+
+			return
+				bx.Cast<MemberAssignment>().OrderBy(b => b.Member.Name).Select((b, i) => new { Expr = b.Expression, b.Member, Index = i })
+				.Join(
+					  by.Cast<MemberAssignment>().OrderBy(b => b.Member.Name).Select((b, i) => new { Expr = b.Expression, b.Member, Index = i }),
+					  o => o.Index, o => o.Index, (xe, ye) => new { XExpr = xe.Expr, XMember = xe.Member, YExpr = ye.Expr, YMember = ye.Member })
+					   .All(o => Equals(o.XMember, o.YMember) && ExpressionsEqual(o.XExpr, o.YExpr, rootX, rootY));
+		}
+
+		private static bool ValuesEqual(object x, object y)
+		{
+			if (ReferenceEquals(x, y))
+				return true;
+			if (x is ICollection && y is ICollection)
+				return CollectionsEqual((ICollection)x, (ICollection)y);
+
+			return Equals(x, y);
+		}
+
+		private static ConstantValue TryCalculateConstant(Expression e)
+		{
+			if (e is ConstantExpression)
+				return new ConstantValue(true, ((ConstantExpression)e).Value);
+			if (e is MemberExpression)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Replace this type-check-and-cast sequence with an 'as' and a null check. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3247)